dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistory.cs

```diff
@@ -124,6 +125,35 @@ public void AddRange(IEnumerable<ChatMessageContent> items)
         this._messages.AddRange(items);
     }
 
+    /// <summary>
+    /// Add a message to the chat history at the end of the streamed message
+    /// </summary>
+    /// <typeparam name="T">Specific chat message content type</typeparam>
+    /// <param name="streamingMessageContents"><see cref="IAsyncEnumerator{T}"/> list of streaming message contents</param>
+    /// <returns>Returns the original streaming results with some message processing</returns>
+    public async IAsyncEnumerable<T> AddStreamingMessageAsync<T>(IAsyncEnumerable<T> streamingMessageContents) where T : StreamingChatMessageContent
+    {
+        List<T> messageContents = new();
+        AuthorRole? role = null;
+        StringBuilder messageContent = new();
+
+        await foreach (var chatMessage in streamingMessageContents.ConfigureAwait(false))
+        {
+            messageContent.Append(chatMessage.Content);
+            messageContents.Add(chatMessage);
+
+            // Is always expected to have at least one chunk with the role provided from a streaming message
+            role ??= chatMessage.Role;
+
+            yield return chatMessage;
+        }
+
+        if (messageContents.Count != 0)
+        {
+            this.AddMessage(new ChatMessageContent(role ?? AuthorRole.Assistant, messageContent.ToString(), messageContents[0].ModelId, messageContents, messageContents[0].Encoding, messageContents[0].Metadata));
```

Comment:
What's in a chunk's metadata? Do we need to somehow combine them? For example, is usage information included in any that we might want to sum?

Reply:
> Is usage information included in any

No, streaming from OpenAI doesn't provide Usage data in the metadata, only, Id, Created and other fields.

Currently the OpenAI `ClientCore` and `WithData` implementations uses the same metadata for all the chunks, this implementation considers this knowledge. I can add a description to make it clear in the code, 

Any fancy implementation to merge multiple dictionaries in here to support other clients that may or may not add different metadata per chunk is something we want? @markwallace-microsoft 
I didn't add this now to not consume extra resources (Less is more).

Reply:
> we might want to sum?

The `Sum` part here can be unpredictable 😊, as each client could provide the same metadata in all chunks, as the example above, or only in the first, and the Sum logic would have to be in theory held by the client.

From a better perspective what can be done here is each `Connector` adds an extension method that handles its special cases, but the method name would have to be different or enforce the `sk developer` to provide an `OpenAIStreamingChatMessageContent`

OpenAI Connector Extension example:
```
public static async IAsyncEnumerable<OpenAIStreamingChatMessageContent> AddMessageContent(this ChatHistory chatHistory, IAsyncEnumerable<OpenAIStreamingChatMessageContent> streaming) {
   // Specific Connector logic
}
```

Reply:
@stephentoub @markwallace-microsoft Implementation moved to the connector, appreaciate a quick review.